### ISSUE TITLE
/graphql error: You don't have permission to access this

### ISSUE THREAD

**Issue by iksent** [directus/directus#4433]
1. Log in as admin
2. Open http://localhost:8055/graphql
3. Get error: 
```
{"errors":[{"message":"You don't have permission to access this.","extensions":{"code":"FORBIDDEN"}}]}
```

---

### Project details
Directus Version: 9.0.0-rc.45
Environment: production
OS: macOS Big Sur (11)
Node: 14.15.4

**Comment by rijkvanzanten** [comment]
GraphiQL under /graphql isn't part of the app, so make sure to "login" to GraphiQL correctly by providing an access token in either the `?access_token=<token>` query parameter, or the `Authorization: Bearer <token>` header 👍🏻

**Comment by MM2-0** [comment]
This works for authenticated users. However, there is currently no way to access the GraphQL API as the public role (neither the GraphiQL interface nor the API itself). A workaround would be to create a new user for the client but since you cannot assign the public role to any user it would mean that you would need to create a new role for that user as well (and copy over all the permissions from the public role).

I am pretty sure that all permissions are set correctly as it did work in rc.44 but the upgrade to rc.45 broke it.

**Comment by rijkvanzanten** [comment]
@MM2-0 if your public role has permissions to read anything, graphql should come to life. If that's not the case, can you open a new issue?

**Comment by MM2-0** [comment]
I found the issue. Apparently the public role now needs read access to directus_fields.field and directus_collections.collection. This wasn't required in rc.44. Maybe that should be stated in the documentation somewhere. Anyway, it is working now.

**Comment by iksent** [comment]
So, to get access to public GraphQL queries from clients we need to make readable `directus_fields.field` and `directus_collections.collection`, right?

**Comment by zanami** [comment]
Same here. That seems rather inconsistent since REST API does not require such permissions.

**Comment by rijkvanzanten** [comment]
@zanami The difference is that the REST pi doesn't expose the available collections and fields, while the GraphQL endpoint does.. the idea was that the collections and fields permissions allow you to control whether or not you want to expose your schema to the world 🤔

**Comment by zanami** [comment]
@rijkvanzanten Still not sure what's the difference. Given the permissions are granted I can do:

```
query {
	collections {
	  collection
	} 
	fields {
	  field
	}
}
```
Why would one need to grant "list **all** collections and fields" permissions to allow for, say, querying Articles collection?
This (directus_fields.field and directus_collections.collection) only exposes collection IDs and field names to graphql, right? Not much new information if so.
Just curious, I don't know much about graphql concepts.

Also, the same permissions allow to browse /collections and /fields which basically exposes the whole data structure in "schema" and "meta".

It actually looks like a bug.

/collections (with only directus_collections.collection perm) exposes this:

```
  {
    collection: 'article',
    meta: {
      collection: 'article',
    },
    schema: {
      name: 'article',
      schema: 'public',
      comment: null,
    },
  },
  {
    collection: 'directus_collections',
    meta: {
      system: true,
      collection: 'directus_collections',
      hidden: false,
      singleton: false,
      icon: 'list_alt',
      note: 'Additional collection configuration and metadata',
      translations: null,
      display_template: null,
    },
    schema: {
      name: 'directus_collections',
      schema: 'public',
      comment: null,
    },
  },
```
Not everything is there for every collection compared to full perms, but for some there is.
Funny thing, 'all access' is granted for Articles, yet there's not much info on it. Just one field is permitted in directus_collections — everything is available.

/collections with 'All access' read permissions on directus_collections
```
{
    collection: 'article',
    meta: {
      collection: 'article',
      icon: null,
      note: null,
      display_template: null,
      hidden: false,
      singleton: false,
      translations: null,
      archive_field: null,
      archive_app_filter: true,
      archive_value: null,
      unarchive_value: null,
      sort_field: null,
    },
    schema: {
      name: 'article',
      schema: 'public',
      comment: null,
    },
  },
  {
    collection: 'directus_collections',
    meta: {
      system: true,
      collection: 'directus_collections',
      hidden: false,
      singleton: false,
      icon: 'list_alt',
      note: 'Additional collection configuration and metadata',
      translations: null,
      display_template: null,
    },
    schema: {
      name: 'directus_collections',
      schema: 'public',
      comment: null,
    },
  },
```
I'm kind of lost, guess I'm missing something.
Fields are even worse. Whatever custom permissions are on directus_fields (tried: field, id, collection) — /fields shows _almost_ everything, I couldn't find much difference from 'All Access':

/fields

```
{
"collection": "article",
"field": "id",
"type": "integer",
"schema": {
"name": "id",
"table": "article",
"data_type": "integer",
"default_value": null,
"max_length": null,
"numeric_precision": 32,
"numeric_scale": 0,
"is_nullable": false,
"is_unique": true,
"is_primary_key": true,
"has_auto_increment": true,
"foreign_key_column": null,
"foreign_key_table": null,
"comment": null,
"schema": "public",
"foreign_key_schema": null
},
"meta": {
"id": 26,
"collection": "article",
"field": "id",
"special": null,
"interface": "numeric",
"options": null,
"display": null,
"display_options": null,
"readonly": true,
"hidden": true,
"sort": null,
"width": "full",
"group": null,
"translations": null,
"note": null
}
},
```

...except for the directus_fields.* other than directus_fields.field, they are not present at all if only 'field' access is granted.

So, if I remove read permission from directus_fields I can't access /fields as a Public user at all. If I grant read permission on directus_fields.field only (Use custom - Field permissions - Field) then I can access full info on every field from all collections _that I have full read permissions to_. I.e. read access to a collection also allows to read everything from directus_fields for that collection, but only if it's allowed to read _anything_ from it.

I'd say something needs to be reworked or at least documented.

**Comment by rijkvanzanten** [comment]
@zanami The idea before was that we didn't want to expose the schema itself (eg the available collections/fields), but one of the main selling points of GraphQL is that it does schema introspection for ease of use and proper errors (eg `did you mean articles instead of artcles?`). Tying it to the same permissions as collections/fields internally meant that you'd have to opt-in to expose your schema, so you won't accidentally leak it. That being said, I agree that it's a little confusing / unfriendly. Besides, the oas spec endpoint returns the same info, so it's time to say that the end user is able to see the collection/field names of the stuff they have access to regardless of permissions to directus_collections/fields

**Comment by rijkvanzanten** [comment]
> I'd say something needs to be reworked or at least documented.

This is already tweaked in https://github.com/directus/directus/pull/4625

**Comment by chrisbartley** [comment]
FWIW, I couldn't get anything from GraphQL via requests from an unauthenticated user (REST API was working fine).   It kept giving me a TypeError with `Definition object should contain 'type' property`.  I'd already modified the Public role for read access to all my collections.  I found this issue, and tried enabling read access to `directus_fields.field` and `directus_collections.collection`, but that didn't change anything.  In frustration, I enabled read access to _all_ directus collections, but that made no difference--same error.  So I turned read access to directus collections back off and decided to try graphql with an authenticated request (from curl)--that worked fine. Great, but doesn't solve my use case for needing read access for unauthenticated users...I figured I'd just use the REST API. 

But then I found, to my total surprise, unauthenticated graphql requests are now working totally fine. I don't understand what the difference could possibly be since I reverted the Public role's permissions back to where they started, but maybe this will help someone :-\
